### PR TITLE
Replace RandomState with Generator

### LINKED ISSUE #166: Replace `RandomState` with `Generator`
#### Is your feature request related to a problem? Please describe
The ALNS library currently uses `numpy.random`'s `RandomState` for random number generation. But this object is considered [legacy](https://numpy.org/doc/stable/reference/random/legacy.html) and no longer receives updates from Numpy 1.16 onwards. 

The main problem happens when users pass their own `rng` to ALNS. The current default RNG from numpy is  [`Generator`](https://numpy.org/doc/stable/reference/random/generator.html), but it has a different interface than `RandomState` (e.g., `integers` vs. `randint`). Since we implement some acceptance criteria/operator selection schemes assuming that the RNG is a `RandomState` object, users that use `Generator` object will run into errors.

#### Describe the solution you'd like
We should replace `RandomState` with `Generator`.

#### Describe alternatives you have considered
N/A

#### Additional context
N/A

### PR DESCRIPTION
**Pull request by leonlan**
This PR:

- [x] Closes #166.
- [ ] Adds and passes relevant tests.
- [ ] Has well-formatted code and documentation.

<details>

**Notes**:

- It is essential that you add a test when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

File: examples/resource_constrained_project_scheduling_problem.ipynb
Comment (leonlan) [inline examples/resource_constrained_project_scheduling_problem.ipynb:1]: https://alns--173.org.readthedocs.build/en/173/examples/resource_constrained_project_scheduling_problem.html

Annoying runtime warning but I can't find the cause.

Comment (leonlan) [reply examples/resource_constrained_project_scheduling_problem.ipynb:1]: Found the issue: indices are initialized with `np.inf`. Initializing with `max(self.jobs)` does the same so we should do that instead.

File: alns/accept/AcceptanceCriterion.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ class AcceptanceCriterion(Protocol):
     """
 
     def __call__(
-        self, rnd: RandomState, best: State, current: State, candidate: State
+        self, rng: Generator, best: State, current: State, candidate: State
```
Comment (N-Wouda) [inline alns/accept/AcceptanceCriterion.py:14]: Docstring (now rng, not rnd).

File: alns/tests/test_result.py [hunk 2]
```diff
@@ -36,15 +36,15 @@ def get_statistics():
         statistics.collect_objective(objective)
 
     # We should make sure these results are reproducible.
-    state = rnd.RandomState(1)
+    state = rnd.default_rng(1)
```
Comment (N-Wouda) [inline alns/tests/test_result.py:39]: rng?

**Review by N-Wouda — APPROVED** [review]
LGTM; few minor things

**Pull request merged**